### ISSUE TITLE
Upgrader doesn't allow fallback behavior

### ISSUE THREAD

**Issue by rwz** [gorilla/websocket#116]
Current implementation doesn't allow to define a fallback behavior for an incoming request that is not websocket.

```go
var upgrader = websocket.Upgrader{}

func (wh *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
  conn, err := upgrader.Upgrade(w, r, nil)

  if err != nil {
    w.WriteHeader(http.StatusTeapot)
    io.WriteString(w, "I'm a tea-pot")
    // can't do that. at this point upgrader has
    // already sent a 400 Bad Request as a response :(
  }
}
```

It'd be nice to have the logic that checks if the incoming request is a proper websocket request to be extracted and be available as a sepatate method.

```go
func (wh *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
  if upgrader.CanUpgrade(r) {
    conn, err := upgrader.Upgrade(r, w)
    // do websocket things
  } else {
    // do http things
  }
}
```

**Comment by garyburd** [comment]
An application can control the HTTP response by setting [Upgrader.Error]((r.Header, "Upgrade", "websocket").

It does not make sense to fallback to HTTP for an invalid websocket handshake.  How about this function instead?

    func IsWebSocketUpgrade(r *http.Request) bool {
        return tokenListContainsValue(r.Header, "Connection", "upgrade") &&
               tokenListContainsValue(r.Header, "Upgrade", "websocket")
    }

**Comment by rwz** [comment]
@garyburd that'd be great. Basically it'd allow to make the same path handle both HTTP and websocket requests.

**Comment by garyburd** [comment]
Fixed by a622679ebd7a3b813862379232f645f8e690e43f
